source/extensions/filters/network/dubbo_proxy/router/router_impl.cc

```diff
@@ -0,0 +1,324 @@
+  if (status == DubboFilters::UpstreamResponseStatus::Complete) {
+    ENVOY_STREAM_LOG(debug, "dubbo router: response complete", *callbacks_);
+    upstream_request_->onResponseComplete();
+    cleanup();
+    return;
+  } else if (status == DubboFilters::UpstreamResponseStatus::Reset) {
+    ENVOY_STREAM_LOG(debug, "dubbo router: upstream reset", *callbacks_);
+    upstream_request_->resetStream();
+    return;
+  }
+
+  if (end_stream) {
+    // Response is incomplete, but no more data is coming.
+    ENVOY_STREAM_LOG(debug, "dubbo router: response underflow", *callbacks_);
+    upstream_request_->onResponseComplete();
+    upstream_request_->onResetStream(
+        Tcp::ConnectionPool::PoolFailureReason::RemoteConnectionFailure);
+    cleanup();
+  }
+}
+
+void Router::onEvent(Network::ConnectionEvent event) {
+  if (!upstream_request_ || upstream_request_->response_complete_) {
+    // Client closed connection after completing response.
+    return;
+  }
+
+  switch (event) {
+  case Network::ConnectionEvent::RemoteClose:
+    upstream_request_->onResetStream(
+        Tcp::ConnectionPool::PoolFailureReason::RemoteConnectionFailure);
+    break;
+  case Network::ConnectionEvent::LocalClose:
+    upstream_request_->onResetStream(
+        Tcp::ConnectionPool::PoolFailureReason::LocalConnectionFailure);
+    break;
+  default:
+    // Connected is consumed by the connection pool.
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+}
+
+const Network::Connection* Router::downstreamConnection() const {
+  return callbacks_ != nullptr ? callbacks_->connection() : nullptr;
+}
+
+void Router::cleanup() {
+  if (upstream_request_) {
+    upstream_request_.reset();
+  }
+}
+
+Router::UpstreamRequest::UpstreamRequest(Router& parent, Tcp::ConnectionPool::Instance& pool,
+                                         MessageMetadataSharedPtr& metadata,
+                                         SerializationType serialization_type,
+                                         ProtocolType protocol_type)
+    : parent_(parent), conn_pool_(pool), metadata_(metadata),
+      deserializer_(
+          NamedDeserializerConfigFactory::getFactory(serialization_type).createDeserializer()),
+      protocol_(NamedProtocolConfigFactory::getFactory(protocol_type).createProtocol()),
+      request_complete_(false), response_started_(false), response_complete_(false) {}
+
+Router::UpstreamRequest::~UpstreamRequest() {}
+
+Network::FilterStatus Router::UpstreamRequest::start() {
+  Tcp::ConnectionPool::Cancellable* handle = conn_pool_.newConnection(*this);
+  if (handle) {
+    // Pause while we wait for a connection.
+    conn_pool_handle_ = handle;
+    return Network::FilterStatus::StopIteration;
+  }
+
+  return Network::FilterStatus::Continue;
+}
+
+void Router::UpstreamRequest::resetStream() {
+  if (conn_pool_handle_) {
+    conn_pool_handle_->cancel(Tcp::ConnectionPool::CancelPolicy::Default);
+  }
+
+  if (conn_data_ != nullptr) {
```

Comment:
ASSERT(! conn_pool_handle_)?

Reply:
conn_pool_handle_ is the processing that the connection pool provides to the upper layer while the connection is being established, and is set to nullptr once the connection is established, the resetStream call may be made after the connection is established

Reply:
ditto

Reply:
done